### PR TITLE
SVCD::subscribe [NOT READY]

### PR DESCRIPTION
**Pull request by eldon**
Implemented subscribe function in C.

### CODE REVIEW

File: natlib/svcd.c [hunk 1]
```diff
@@ -185,3 +186,50 @@ static int svcd_init( lua_State *L )
...
+    ((uint16_t*) (msg+1))[0] = luaL_checknumber(L, 2); //svcid
+    ((uint16_t*) (msg+1))[1] = luaL_checknumber(L, 3); //attrid
+    ((uint16_t*) (msg+1))[2] = ivkid;
+    lua_pushlstring(L, msg, 7);
+     
+    lua_pushvalue(L, 1);
+    lua_pushnumber(L, 2530);
+    lua_call(L, 4, 0);
+    lua_pushnumber(L, ivkid);
+    return 0;
```
Comment (leonardt) [inline natlib/svcd.c:234]: You should return the number of results (1 in this case).

men in #000000

**Comment by jenniferdai** [comment]
lua_getfield(L, -1, "ivkid");
I might be misunderstanding, but I thought L, -1 gets the first parameter, but ivkid is the second parameter?

-team birdpeople

**Comment by mccormickt12** [comment]
@jenniferdai the -1 is referring to the placement of the table being indexed to by "ivkid", so it should be fine

-TBD

File: natlib/svcd.c [hunk 2]
```diff
@@ -185,3 +186,50 @@ static int svcd_init( lua_State *L )
...
+    msg[0] = 1;
+    ((uint16_t*) (msg+1))[0] = luaL_checknumber(L, 2); //svcid
+    ((uint16_t*) (msg+1))[1] = luaL_checknumber(L, 3); //attrid
+    ((uint16_t*) (msg+1))[2] = ivkid;
+    lua_pushlstring(L, msg, 7);
+     
+    lua_pushvalue(L, 1);
+    lua_pushnumber(L, 2530);
+    lua_call(L, 4, 0);
+    lua_pushnumber(L, ivkid);
```
Comment (pnigam) [inline natlib/svcd.c]: We aren't sure why you are pushing ivkid here. Looking at the lua code, this doesn't seem necessary.

File: natlib/svcd.c (hunk 1 above)
Comment (tbehailu) [inline natlib/svcd.c:234]: Looks like you removed SVCD before constructing your message and sending it & wasn't sure if you could do that.

Sincerely, Running with Scissors

File: natlib/svcd.c [hunk 3]
```diff
@@ -185,3 +186,50 @@ static int svcd_init( lua_State *L )
...
+
+static int svcd_subscribe(lua_State* L)
+{
+    if (lua_gettop(L) != 4) return luaL_error(L, "Expected (targetip, svcid, attrid, on_notify)");
+     
+    lua_getglobal(L, "SVCD");
+    lua_getfield(L, -1, "ivkid");
+     
+    uint16_t ivkid = luaL_checkinteger(L, -1);
+    lua_pop(L, 1);
```
Comment (jochemvang) [inline natlib/svcd.c:198]: Nit-picking, but I don't think you need to do this per se. Can't you just leave this on the stack and have it cleaned up later?
- PB for your Jelly or, The Perfect Storm

File: natlib/svcd.c [hunk 4]
```diff
@@ -185,3 +186,50 @@ static int svcd_init( lua_State *L )
...
+    // put value on the stack
+    lua_pushinteger(L, ivkid);
+    lua_pushvalue(L, 4); // on_notify
+    lua_settable(L, -3);
+    lua_pop(L, 2);
+     
+     
+    lua_pushlightfunction(L, libstorm_net_sendto);
+     
+    lua_getglobal(L, "SVCD");
```
Comment (hsysuper) [inline natlib/svcd.c:218]: You have already gotten SVCD in the beginning of the function. If you know that you might need it again, why not just keeping it in the memory and recycle it here?

Untitled

File: natlib/svcd.c [hunk 5]
```diff
@@ -185,3 +186,50 @@ static int svcd_init( lua_State *L )
 
     return 0;
 }
+
+static int svcd_subscribe(lua_State* L)
+{
+    if (lua_gettop(L) != 4) return luaL_error(L, "Expected (targetip, svcid, attrid, on_notify)");
+     
+    lua_getglobal(L, "SVCD");
+    lua_getfield(L, -1, "ivkid");
```
Comment (immesys) [inline natlib/svcd.c:195]: Good job! much nicer function than gettable()

File: natlib/svcd.c [hunk 6]
```diff
@@ -185,3 +186,50 @@ static int svcd_init( lua_State *L )
...
+     
+    uint16_t ivkid = luaL_checkinteger(L, -1);
```
Comment (immesys) [inline natlib/svcd.c:197]: checkinteger is really only meant to be used for parameters. I think you probably want lua_tointeger, or lua_tonumber

File: natlib/svcd.c [hunk 7]
```diff
@@ -185,3 +186,50 @@ static int svcd_init( lua_State *L )
...
+    lua_pushlightfunction(L, libstorm_net_sendto);
+     
+    lua_getglobal(L, "SVCD");
+    lua_getfield(L, -1, "ncsock");
+    lua_remove(L, -2); // remove SVCD
+     
+     
+    uint8_t msg[7];
+    msg[0] = 1;
+    ((uint16_t*) (msg+1))[0] = luaL_checknumber(L, 2); //svcid
```
Comment (immesys) [inline natlib/svcd.c:225]: definitely do not want to use checknumber here. use tonumber()

Comment (immesys) [inline natlib/svcd.c:225]: this is incorrect code for ARM. You are attempting to write a uint16_t to an unaligned section of an unaligned array. At best, the compiler will happen to emit instructions that can support unaligned writes, at worst you will crash the processor with a Hard Fault.

Memory alignment is a subtle problem, so don't feel bad. You have a couple possible solutions:
a) do everything as bytes
b) allocate msg as an array of 8 bytes, with:

```
 __attribute__((aligned(2)))
```

ignore the first byte, set the second byte to 1 and then write your uint16_t's as you have done - this way they will be aligned
c) declare a packed struct and use that to populate your message. Note you will need to manually pad an extra byte in the front, same as in b

File: natlib/svcd.c (hunk 1 above)
Comment (nikitakit) [inline natlib/svcd.c:234]: Fixed return value

File: natlib/svcd.c (hunk 6 above)
Comment (nikitakit) [inline natlib/svcd.c:197]: fixed

File: natlib/svcd.c (hunk 7 above)
Comment (nikitakit) [inline natlib/svcd.c:225]: checknumber fixed

File: natlib/svcd.c (hunk 3 above)
Comment (nikitakit) [inline natlib/svcd.c:198]: Changed to stack cleanup to the lua call machinery

File: natlib/svcd.c (hunk 4 above)
Comment (nikitakit) [inline natlib/svcd.c:218]: Changed to stack cleanup to the lua call machinery

File: natlib/svcd.c (hunk 7 above)
Comment (nikitakit) [inline natlib/svcd.c:225]: alignment issues addressed

**Comment by nikitakit** [comment]
See #45 for the revised version

**Comment by immesys** [comment]
superseded by #45

**Pull request closed without merging**